Buildcache Views

kwryankrattiger opened the pull request:
Enable generating index files for build caches containing a subset of specs.

The primary motivation for this change is described in this [discussion](https://github.com/spack/spack/discussions/45629).

This feature is enabled by #48713 which allow concurrent writing of duplicate specs using the build cache layout v3. While likely possible to implement, views are currently not made available for OCI caches.

Build cache views allow for stack/distribution curation to happen using the same unified base cache, while generating "view" indices containing only the packages for each specific distribution. This reduces the total footprint required to provide curated build caches.

For the Spack build caches, each curated stack (ie. HEP/Data-Vis SDK/Radiuss/E4S/etc.) has been required to maintain it's own build cache, and then "promote" binaries from each build cache to a top level cache. This resulted in duplication not only between stacks, but also with the top level cache.

This problem was further exacerbated with snapshot mirrors for develop and releases. These represented full copies of of all binaries concretized in a pipeline. Under the new layout, snapshot storage cost will be reduced from terabytes to megabytes. This will make it possible to persist snapshots longer and/or provide more frequent snapshots (ie. nightly/semi-weekly).

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

kwryankrattiger committed:
Buildcache Views

Enable generating index files for build caches containing a subset of
specs.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Update completions

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

mvandenburgh commented on lib/spack/spack/mirrors/mirror.py:
```diff
@@ -304,6 +323,9 @@ def get_url(self, direction: str) -> str:
 
         return _url_or_path_to_url(url)
 
+    def get_view(self, direction: str):
```
```suggestion
    def get_view(self, direction: Literal["push", "fetch"]):
```

kwryankrattiger replied:
This can also come from a user input so I don't know if the literal would work here. We _could_ make it an enum type to move the validation out.

marcmengel reviewed (approved):
Looks good; has tests and fish command completion flags.
Didn't see any obvious errors.

haampie commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -797,6 +828,102 @@ def update_index_fn(args):
+            if "spec" in data:
+                # This is a spec file, return the hashes of all of the nodes
+                tty.debug(f"Reading {source} as specfile")
+                return [n["hash"] for n in data["spec"]["nodes"]]
```
Make use of SpecReader. There are various versions of the file format.

haampie commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -797,6 +828,102 @@ def update_index_fn(args):
+            elif "concrete_specs" in data:
+                # This is a lock file
+                tty.debug(f"Reading {source} as lockfile")
+                return [h for h in data["concrete_specs"]]
```
Also here, use existing functionality to load the file and then query with public API.

haampie commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -797,6 +828,102 @@ def update_index_fn(args):
+    """Read all of the concrete hashes from a given source"""
+
+    if re.match(r"^/[^ ]{32}$", source):
+        return [source[1:]]
```
This looks a bit curious to me, it's the only Spack command that would support full hashes only (and doesn't use a spec parser). Also, it doesn't validate input properly.

I would suggest to remove this branch. As far as I can see, it's pretty useless cause users have to list all dependencies themselves.

kwryankrattiger replied:
Ah yeah, that will simplify this a fair amount!

haampie commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -797,6 +828,102 @@ def update_index_fn(args):
+            # Load the current state of the view index from the cache into the database
+            cache_index = BINARY_INDEX._local_index_cache[str(url_and_version)]
+            cache_key = cache_index["index_path"]
+            db._read_from_file(BINARY_INDEX._index_file_cache.cache_path(cache_key))
```
Looks like you're reading this into memory but never save the database.

It would make more sense if you either copied the existing database or created a new one.

Also, append should append to the upstream, not the possibly outdated local download of the view. Checking the index whether it's up to date should be cheap.

kwryankrattiger replied:
> Looks like you're reading this into memory but never save the database.

The database is a temporary to be used to generate the index and thrown away

> It would make more sense if you either copied the existing database or created a new one.

We are initializing the DB for the mirror index we are generating with the "current" state of the remote mirror for append. 

> Also, append should append to the upstream, not the possibly outdated local download of the view. Checking the index whether it's up to date should be cheap.

We do our best on this, but it is impossible to lock the remote index to ensure perfect consistency. There is a warning that is printed whenever `--append` is used stating that the behavior is racy.

The first thing we do in this command is `BINARY_INDEX._fetch_and_cache_index(url_and_version)` to force update the local index. This supersedes any of the time stamp checking used by the normal update call, which ensures that I am getting the latest state of the view index locally.

haampie commented on lib/spack/spack/cmd/commands.py:
Please revert these unrelated changes, you can always submit a separate PR.

kwryankrattiger replied:
Ah, yes, I had been meaning to do that and forgot about it.

haampie commented on lib/spack/spack/mirrors/mirror.py:
```diff
@@ -158,6 +168,15 @@ def ensure_mirror_usable(self, direction: str = "push"):
+        if binary_layout_version == 2:
+            return self.get_view("fetch") is not None
+
+        return True
```
I've read this twice and still don't understand what it's supposed to do. Looks either a naming issue, or an invariant that should be enforced by the class instead of by the call site.

kwryankrattiger replied:
basically it is testing if the build cache layout version supports a configured feature in the mirror. Views are not supported in v2 mirrors, and there will likely be other features we add on top of v3 that will also not be supported.

I am open to alternative names.

haampie commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -771,11 +803,14 @@ def _url_generate_package_index(url: str, tmpdir: str):
         try:
-            _read_specs_and_push_index(file_list, read_fn, url, db, str(db.database_directory))
+            _read_specs_and_push_index(
+                file_list, read_fn, name, filter_fn, url, db, str(db.database_directory)
```
Looks like you're passing `filter_fn` to the wrong function. `get_entries_from_cache` downloads all spec files from the build cache, that's exactly what you're trying to avoid.

kwryankrattiger replied:
In the case of using the `awscli` tool, we do an s3 sync of manifests to a local cache. This is an "optimization" in the general case. I am not planning on changing this here since it would require more investigation into heuristics around when it is better to sync everything with aws sync vs pull manifests lazily. Deeper investigation can come in a follow up.

In all other cases they are fetched lazily.

We don't actually read the spec.json blobs until we call `_read_specs_and_push_index` after checking if the spec should be included or not.

```
spec_dict = cache_entry.fetch_metadata()
```

haampie commented on lib/spack/spack/url_buildcache.py:
```diff
@@ -1335,25 +1336,35 @@ class MirrorURLAndVersion:
-            return self.url == other.url and self.version == other.version
+            return all(
+                [self.url == other.url, self.version == other.version, self.view == other.view]
+            )
```
Please don't allocate an array to compare three properties.


```suggestion
            return self.url == other.url and self.version == other.version and self.view == other.view
```

haampie commented on lib/spack/spack/url_buildcache.py:
```diff
@@ -1335,25 +1336,35 @@ class MirrorURLAndVersion:
-        return f"{self.url}__v{self.version}"
+        s = f"{self.url}__v{self.version}"
+        if self.view:
+            s += f"__@{self.view}"
```
Why do you need both `@` and `__`? Can't split on `__` anyways.

kwryankrattiger replied:
To be consistent with `__<key>`. Syntactically it doesn't really matter. If I add a restriction to the view names s.t. they can start with a number we could drop it completely and remove the extra match group for handling the conditional delimiter.

haampie commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -438,16 +438,20 @@ def update(self, with_cooldown=False):
                         needs_regen = self._fetch_and_cache_index(
                             urlAndVersion, cache_entry=cache_entry
                         )
                         self._last_fetch_times[urlAndVersion] = (now, True)
                         all_methods_failed = False
                     except FetchIndexError as e:
-                        needs_regen = False
                         fetch_errors.append(e)
                         self._last_fetch_times[urlAndVersion] = (now, False)
+                    except BuildcacheIndexNotExists:
+                        self._last_fetch_times[urlAndVersion] = (now, False)
+                        all_methods_failed = False
```
Can you leave some comments why these cases are handled differently?

kwryankrattiger replied:
Sure

haampie commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -438,16 +438,20 @@ def update(self, with_cooldown=False):
                         all_methods_failed = False
                 else:
                     # May need to fetch the index and update the local caches
```
In this function you don't initialize `configured_mirrors` correctly, it's lacking the view name.

Please fix and add a test.

I think you should also test that whenever a user adds, changes or removes the view in config, we don't continue to fetch the old index. It would be good to make that a unit tests instead of an integration test: structure the code so you can test what `items_to_remove` is.

kwryankrattiger replied:
Ah yeah, bad rebase after your fixes to the OCI.

I agree this section needs more testing, but I would like to avoid taking on the larger task of refactoring the cache so we can test it better. The existing tests just add a binary mirror with an index and check that the expected specs are listed with `spack buildcache list -la`. I will do that for a mirror configure with a view to at least have the same level of testing as the non-view version.

haampie commented on lib/spack/spack/url_buildcache.py:
```diff
@@ -1335,25 +1336,35 @@ class MirrorURLAndVersion:
     def from_string(cls, s: str):
-        parts = s.split("__v")
-        return cls(parts[0], int(parts[1]))
+        m = re.match(r"^(.*)__v([0-9]+)(__@(.*))?$", s)
```
Use non-capturing groups for things you don't need

```suggestion
        m = re.match(r"^(.*)__v([0-9]+)(?:__@(.*))?$", s)
```

kwryankrattiger replied:
Neat, didn't know that existed!

kwryankrattiger committed:
Address review comments

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

psakievich commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -848,3 +850,192 @@ def test_buildcache_prune_orphaned_manifest(tmp_path, mutable_database, mock_gnu
+    mpileaks_1_hashes = [s.dag_hash() for s in mpileaks_specs[1].traverse()]
+    mpileaks_2_hashes = [s.dag_hash() for s in mpileaks_specs[2].traverse()]
+
+    # Make sure the view doesn't exist yet
+    with pytest.raises(spack.binary_distribution.FetchIndexError):
+        hashes_in_view = read_specs_in_index(mirror_directory, "test_view")
```
I'm seeing this repeated. I'm wondering if there is a need for a user interface to check if a view exists? Say I don't have write access, only read to the mirror? If that's scope creep let me know. I won't hold up an approval over this. I only see a pattern and wonder if it is going to be accounted for in the user API.

kwryankrattiger replied:
I agree it probably makes sense to have a way to check if the configuration of a mirror is going to do what the user wants. I kind of have a check that the version of the layout supports features in a mirror, but that is also hidden.

It has been on my wish list to have a `spack mirror check <mirrors...>` which would do a more indepth check of the mirror and report things like orphaned blobs, holes in dependencies, feature compatibility with config, deprecated layouts, signatures match match with keys in key index, etc. with options to run the different checks. Running these types of checks hit the network, so I don't really want to run them unless the user wants them.

becker33 replied:
I agree with @psakievich, this needs to be some sort of API call instead of a failing command to test it.

kwryankrattiger replied:
tl;dr - I agree, I think it can be a follow on. The check here is a quick and dirty check that I added for the tests only, and only to verify the initial state of the view to make sure the test is starting from a known state.

I don't think this is generally going to be something people need to check, it is more a nice to have when working with views. If the view index does not exist the fall back for the mirror will be to use direct fetch only just like the case where the cache index does not exist. Like the top level index, there is no requirement that a view index needs to exist to use a mirror configured with a view.

The use here is to make sure the state of the test is as expected to avoid a case where we are testing something that is not the actual state.

I have a a few build cache checks that I am planning to translate out of the build cache tools that we maintain in infra. The main two there are to validate an index and  to verify that the blobs referenced in manifests all exist. Adding a check for a view index existing if a view is configured would be an easy thing to add in addition to that.

psakievich commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -848,3 +850,192 @@ def test_buildcache_prune_orphaned_manifest(tmp_path, mutable_database, mock_gnu
+
+    # Test failure due to existing index
+
+    # Test append
```
Seems like these should be separate tests rather than one super test. It will make it easier to debug failures in the future. Could nest them in a test class.

kwryankrattiger replied:
I have them all in the same test to test the commands against different states of a view index to ensure the command is doing what I expect for the options. I could break it up into more though.

kwryankrattiger replied:
I broke the test into three x ways to pass a spec

* `spack update-view` to create a view and then makes sure that `--force` is required overwrite an existing view.

* `spack update-view --append ...`. I found a bug writing this one where using `--append` when an index didn't exist it would error (fixed). So this test tests the `--append` to a non-existent view creates the view, and then appending to an existing view (created by the above command) successfully appends.

* `spack update-view --force ` can overwrite an existing view

psakievich commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -848,3 +850,192 @@ def test_buildcache_prune_orphaned_manifest(tmp_path, mutable_database, mock_gnu
+    for h in mpileaks_0_hashes + mpileaks_1_hashes:
+        assert h in hashes_in_view
+
+    # Test fail to overwrite
```
Seems like this comment is stale and you are testing intersection of the view vs the full cache? Would be clearer as a separate test.

kwryankrattiger replied:
Yeah, too much coyp-paste here. Fixed.

psakievich commented on lib/spack/spack/test/binary_distribution.py:
```diff
@@ -1317,3 +1344,41 @@ def urlopen(request: urllib.request.Request):
+    from_str = spack.binary_distribution.MirrorURLAndVersion.from_string(as_str)
+
+    # Verify round trip
+    assert url_and_version.url == "https://dummy.io/__v3__@aview"
```
This URL feels really awkward in that it isn't a valid url any more correct? Do we need to encode the logic in the url? Is there a reason to not nest these in config in some form like below?

```
url: https://dummy.io
version: 3
views:
- aview
```

If there is a compelling reason why the view has to expressible as a single URL then this implementation seems as good as any. But when I look at the object below it is all split. So URL is feeling overloaded to me in this review. Which is it? `url = path+version+view` or `url=path`

kwryankrattiger replied:
The reason is around how we cache indices. This class is a key into the cache index.

Even if we allowed configuring multiple views per mirror, which we could, internally we would still want to cache each index by URL/version/view to avoid complexity in the fetchers and ensuring we aren't fetching the same index over and over.

Imo, the utility of having multiple views per mirror isn't enough for me to want to add the complexity. If others disagree though I am open to it.

psakievich replied:
That makes sense to me. I think my real criticism is using the key `url`.  It is counterintuitive if it is not a real url. I would consider renaming that if we can in this PR to something like `key` or `identifier`. Really it is `url+version+view` so I'd prefer to not misdirect by preserving the same key. 

It is subtlety that is bordering on bug for end users since when you aren't using a view, `url == url` but when you are using a view `url != url`. The config representation was more about decomposing the key rather than indexing a list of views. That was really just a last minute bonus idea.

kwryankrattiger replied:
Ah, I see why this is confusing. So that URL really is the URL. What I was testing here is if a URL contains the delimiters used by the `MirrorUrlAndVersion` serialization can I correctly recover the URL, version, and view. I had an issue with using `split` and and regex patterns around the delimiters and if they weren't constructed correctly it would fail in this case.

psakievich commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -2202,6 +2244,12 @@ def try_direct_fetch(spec, mirrors=None):
+            if not mirror.supported_version(binary_layout_version=layout_version):
+                # Check if the features of this mirror are supported by the layout version
+                # ie. only v3 mirrors have view support
+                continue
```
I find this a little confusing. So we are validating the mirror supports the features of the current layout version before proceeding, but then the calling code after is `layout_version` independent? What happens if the mirror has a feature that can be used in v2 and v3? Does it pull from the mirror twice and give duplicates in `found_specs`?  

I also noted @haampie commented on the `supported_version` function above. I agree with his critique and I think the flow control here is why it is confusing. I think we should embed this logic into the functions that actually use the `layout_version` i.e. `get_url_buildcache_class` and `MirrorURLAndVersion`.

IMO control on features that are supported with versioning should live closer to the implementation of the features. This limits the chance for logic errors on these more abstract methods that span all the layouts.

kwryankrattiger replied:
Yeah, being two for two on this being confusing is a sure sign it needs reworking. I will think about how to make this cleaner.

It is kind of an awkward thing because it is basically saying "don't try to do stuff with V2 Layout If a View is configured because it won't work". But as you have pointed out, this isn't the level that actually _uses_ the layout or the view.

kwryankrattiger replied:
So we need to know this information from the mirror.

What I think I am going to do is convert all uses of `SUPPORTED_LAYOUT_VERSIONS` to `mirror.supported_layout_versions`. This way a mirror says what is supported for it rather than trying everything that could be possible.

psakievich reviewed (commented):
Trying to spin up on all of this. Mostly looks good, but I do have some concerns around the url specification/overloading which I'd like to address before going much deeper. I primarily focused my review around the tests since I don't know the guts of `buildcache.py` very well.

kwryankrattiger committed:
Split create and append tests. Fix append bug

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Use the mirror to get supported layout versions

Move SUPPORTED_LAYOUT_VERSIONS and `signed` checking logic out of
inlined checks in the `spack.binary_distribution` module into a
function in the `spack.mirrors.mirror.Mirror` class. This consolidates
the logic about what features are available in a mirror to a single
place rather than having to remember to implement the checks everywhere
they are needed.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Add docs

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

haampie commented on lib/spack/spack/url_buildcache.py:
```diff
@@ -1335,25 +1333,35 @@ class MirrorURLAndVersion:
+    def __init__(self, url: str, version: int, view: Optional[str] = None):
         self.url = url
         self.version = version
+        self.view = view
```
I think this is not right. These objects are used in `BinaryCacheIndex` to look up all mirrors containing a certain spec. For fetching from the binary cache the view is redundant. When constructing a list of unique mirrors that contain a spec, the view should not play a role (currently it does with `__eq__` and `__hash__`).

kwryankrattiger replied:
Not necessarily true, it is possible to have a build cache with no index other than the view index(es). In that case, each view will provide a a different index (potentially non-overlapping). So we have to fetch and cache them like they are unique mirrors, because as far as Spack is concerned they kind of are.

haampie reviewed (dismissed):
See the comment above.

kwryankrattiger committed:
Move update-view under existing update-index

The view and index are the same, views just have a name. This way
existing workflows with update-index can work the same with views
without having an entirely new syntax.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Fixes post rebase

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
MirrorForSpec list should strip the view

The view is not important for building the URL for fetching. Strip it
from the mirror info used for building the fetch URL.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Update completions

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

zackgalbreath commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,9 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+
+Build caches can quickly become large and inefficient to search as binaries are added over time.
+A common work around to this problem is to break the build cache into stacks that target specific applications or workflows.
+This allows for curation of binaries as smaller collections of packages that push a their own mirrors that each maintain a smaller search area.
```
```suggestion
This allows for curation of binaries as smaller collections of packages that push to their own mirrors that each maintain a smaller search area.
```

zackgalbreath commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,9 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+
+
+To better address the issues with large search areas, build cache index views (or just "views" in this section) were introduced.
+A view is a named index which provides a curated view into a larger buildcache.
```
```suggestion
A view is a named index which provides a curated view into a larger build cache.
```

zackgalbreath commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,9 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+A view is a named index which provides a curated view into a larger buildcache.
+This allows build cache maintainers to provide the same granularity of build caches split by stacks without having to pay for the extra storage and compute required for the duplicated dependencies.
+
+Views can be created or updated using an active environment, environment lockfile(s), specfile(s), or specifying a spec hash(es) indexed in the local database.
```
```suggestion
Views can be created or updated using an active environment, environment lockfile(s), specfile(s), or by specifying one or more spec hashes indexed in the local database.
```

zackgalbreath commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,9 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+
+Views can be created or updated using an active environment, environment lockfile(s), specfile(s), or specifying a spec hash(es) indexed in the local database.
+
+View indices are stored similarly to the top level build cache index, but use an additional prefix of the view name ``<build cache prefix>/v3/manifests/index/my-stack-view/index.manifest.json``.
```
```suggestion
View indices are stored similarly to the top-level build cache index, but use an additional prefix of the view name ``<build cache prefix>/v3/manifests/index/my-view-name/index.manifest.json``.
```

zackgalbreath commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,9 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+``spack buildcache create-view``
+""""""""""""""""""""""""""""""""
+
+Here is an example of creating a view using from an active environent.
```
```suggestion
Here is an example of creating a view using an active environent.
```

zackgalbreath commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,9 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+``spack buildcache update-view``
+""""""""""""""""""""""""""""""""
+
+Updating the existing view using appending as the stack moves forward.
```
```suggestion
```

zackgalbreath commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,9 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+""""""""""""""""""""""""""""""""
+
+Updating the existing view using appending as the stack moves forward.
+To prevent accidently overwriting an existing view it is required to specify how a view should be handled by update.
```
```suggestion
To prevent accidently overwriting an existing view, it is required to specify how a view should be updated.
```

zackgalbreath commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,9 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+To prevent accidently overwriting an existing view it is required to specify how a view should be handled by update.
+The two options for handling updates are ``--force`` and ``--append``.
+Using the force option will replace the index as if the previous one did not exist.
+The append option will first read the existing index, and then add the specs specified the command.
```
```suggestion
The append option will first read the existing index, and then add new specs to it.
```

zackgalbreath commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,9 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+
+.. warning::
+
+   Using the ``--append`` option with build cache index views is a non-synchonous operation.
```
```suggestion
   Using the ``--append`` option with build cache index views is a non-atomic operation.
```

zackgalbreath commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -866,4 +1035,9 @@ def prune_fn(args):
 
 def buildcache(parser, args):
+    try:
+        # update-index supports unkwown_args for view extension
```
```suggestion
        # update-index supports unknown_args for view extension
```

zackgalbreath commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -999,3 +1000,259 @@ def mock_stat_url(url: str):
+    empty_lockfile.write_text(
+        '{"_meta": {"lockfile-version": 1}, "roots": [], "concrete_specs": {}}', encoding="utf-8"
+    )
+    # Create a view with now specs
```
```suggestion
    # Create a view with no specs
```

zackgalbreath reviewed (dismissed):
Suggested typo/grammar cleanups, otherwise the proposed code changes look reasonable to me

kwryankrattiger committed:
Various fixes to docs

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

becker33 commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,10 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+
+   Using the ``--append`` option with build cache index views is a non-atomic operation.
+   In the case where multiple writers are appending to the same view, the result will only include the state of the last to write.
+   When using ``--append`` for build cache workflows it is up to the user to correctly serialize the update operations.
```
I don't think it's fair to single out the `--append` workflow for this exactly -- every possible workflow using `--force` also shares the same raciness, it's just slightly more obvious.

becker33 commented on lib/spack/docs/binary_caches.rst:
```diff
@@ -137,10 +137,70 @@ For example, to combine all of the commands above to add the E4S build cache and
+
+To better address the issues with large search areas, build cache index views (or just "views" in this section) were introduced.
+A view is a named index which provides a curated view into a larger build cache.
+This allows build cache maintainers to provide the same granularity of build caches split by stacks without having to pay for the extra storage and compute required for the duplicated dependencies.
```
We should note here that OCI buildcaches do not support views

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +820,157 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+    print(f"mode: {update_mode}")
+    print(f"name: {name}")
+    for source in sources:
+        print(f"source: {source}")
```
This looks like vestigial debugging code. If you want to keep it, we should turn it into a `tty.msg` or `tty.debug` call.

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +820,157 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+        )
+
+    hashes = []
+    if sources:
```
I think it might be preferable to create separate arguments for the different types of sources here, because that would allow us to reference an environment by name as a source (otherwise that's ambiguous with a spec).

It would also be nice if we could list specs by constraints, not just by hash. E.g. `spack buildcache update-view --append %gcc@14.3.0` to add all specs built with that compiler to the view. It might need a `--all` option as well to make that make sense.

kwryankrattiger replied:
I am opposed to supporting the spec filtering in this command directly because the point of the view is it is explicitly adding specs from a concrete source. So those will either come from an environment (lock), database (index), or from a single concrete spec hash. How and where those sources are discovered from is left to the user as it is, imo, beyond the scope of this command to have more advanced searching functionality.

You can achieve the same thing by doing this, which builds off of the `spack find` command. This seems to me more ideal as `spack find` is already the "advanced search" command, and features and bug fixes around that functionality will be applied directly to it.

```
spack buildcache update-index my-mirror-with-view $(spack find --format {/hash} "%c=gcc@14.3.0")
```

becker33 replied:
We generally have specs as the input, not hashes, because specs are our general abstraction for user input. I don't think we should deviate from that here, and I'm not (at least yet) seeing any negative to doing it. What do we lose by filtering on specs here?

There isn't an issue with code duplication, we have a very nicely defined interface to query specs (`Database.query`) which is the same entrypoint used by `spack find`.

kwryankrattiger replied:
My only worry here is still that the user may pass a spec and not include specs that are in the cache, but not installed locally. `spack buildcache update-index` has historically worked anywhere. In order to do this reliably, we would have to first (potentially create/update) sync the top level index, then query they database, and then update the view index. \

Updating the top level index can be expensive and having that as a requirement is not great. It is also not necessarily true that there will/should be a top level index if view indices are being used. I would rather have the command provide an interface that lends itself to predictable behavior.

The other option I see here is to just drop the spec list feature all together for now can revisit it later when a more concrete (pun) use case is presented. For all of the purposes I have in mind for this, view indices will be derived exclusively from lockfiles/concrete environments.

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +820,157 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+
+    # Try to read input as a spec
+    try:
+        concrete_specs = spack.cmd.parse_specs(source, concretize=True)
```
I think this should be searching for installed specs, rather than just concretizing the given inputs

kwryankrattiger replied:
ah, this is maybe my misunderstanding then. I thought it was required to pass `concretize=True` to get a concrete spec back.

I didn't query the db directly since it the goal was to support short hashes as well. We also want to support including specs that are not installed, but may exist in the remote cache. It is completely reasonable in this case to list uninstalled concrete specs to include in the index of a build cache. We check later if that spec is in the cache, and mark in `in_buildcache: true|false` appropriately.

The only hard requirement is the hash must exist in the local database so we can get back a concrete spec object.

becker33 replied:
The problem with concretizing is that you could pass it a constraint for which there is no matching spec already known to spack.

It sounds like you're trying to enforce that the input specs include a hash? Then I think what you want is something like:
```
abstract_specs = spack.cmd.parse_specs(source)
for s in abstract_specs:
    if not s.abstract_hash:
        raise SomeSortOfException("message")

concrete_specs = [s.lookup_hash() for s in abstract_specs]
```

But we could also require something more complicated like "each input spec matches either a unique installed spec or has an abstract hash", in which case we could do something like:
```
abstract_specs = spack.cmd.parse_specs(source)
concrete_specs = []
for s in abstract_specs:
    if s.abstract_hash:
        concrete_specs.append(s.lookup_hash())
        continue

    try:
        spack.store.db.query_one(s)
    except AssertionError:            
        raise SomeSortOfException("message")
```
Or we could roll our own error checking here for queries that return more than one spec (could be combined with a `--all` option to allow all of them as well, which might be preferable anyway).

becker33 commented on lib/spack/spack/url_buildcache.py:
```diff
@@ -1360,58 +1358,63 @@ class MirrorURLAndVersion:
     in some layout version."""
 
-    __slots__ = ("url", "version")
+    __slots__ = ("url", "version", "view")
```
This class should probably be renamed to something like `MirrorMetadata` now that it's expanding its scope

kwryankrattiger replied:
Fair, I can do that.

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -975,3 +976,259 @@ def mock_stat_url(url: str):
+        hashes_in_view = read_specs_in_index(mirror_directory, "test_view")
+
+    # Create a view using a parametrized source that contains one of the mpileaks
+    with globals()[f"spec_{sources}"](mpileaks_specs[0], tmp_path / "source_stage") as source_args:
```
The only place we use the `sources` string is in this lookup, can we change this to:

```suggestion
    with sources(mpileaks_specs[0], tmp_path / "source_stage") as source_args:
```
And then change the parametrization to
```
@pytest.mark.parametrize("sources", (spec_hash, spec_specfile, spec_env, spec_lockfile))
```
And the same for all of the analogous tests?

becker33 replied:
Actually, I think we could do one better than that. We can make a single pytest fixture that inherits from `mutable_config`, `mutable_mock_env_path`, `tmpdir_factory`, and `requests`. Then it should be able to:

1. Only concretize the environment once
2. Remove the path as an argument (we only ever pass one value)
3. Check the `requests` fixture to know which type of source to return
4. Return a callable context manager that accepts a spec and does the appropriate thing in each case.

kwryankrattiger replied:
I need to read this a couple more times to understand what you mean.

kwryankrattiger replied:
I just went with the first option you gave here, easier for me to wrap my head around.

kwryankrattiger committed:
Rename mirror metadata struct to be more generic

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Add command to check if an index exists

Also add related build cache checks for blobs and manifests.

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Silence very noisy output from fetchers

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
No globals + f-strings :rocket:

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Only support index view create from environment(s)

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Doc formatting

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+        env = ev.environment_from_name_or_dir(source)
+        return list(env.all_hashes())
+    except ev.SpackEnvironmentError as e:
+        raise spack.error.SpackError(f"Failed to read specs from source: {source}") from e
```
This seems to be overengineered now that we've reduced the scope to only using environments as sources. I think we should just replace this with code inline in the caller 

```
    env = ev.environment_from_name_or_dir(source)
    env.all_hashes()  # is already a list
```

and delegate throwing the error to the `spack.environment` module.

kwryankrattiger replied:
I didn't defer the exception on purpose, the name of the source and the error related to it may not be clear alone.

On the list(), I thought I removed that..will fix

becker33 replied:
The error is very clear here:

```
$ spack buildcache update-view --append --name my-view my-mirror my-env1 my-env2
==> Error: no such environment 'my-env2'
```
We shouldn't introduce an indirection layer here where we don't need it.

kwryankrattiger replied:
yeah that is fine for now. I dropped the function completely then since the extra error handling was the only thing that really mattered there.

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+
+    url_and_version = spack.binary_distribution.MirrorMetadata(
+        url, spack.binary_distribution.CURRENT_BUILD_CACHE_LAYOUT_VERSION, name
+    )
```
```suggestion
    metadata = spack.binary_distribution.MirrorMetadata(
        url, spack.binary_distribution.CURRENT_BUILD_CACHE_LAYOUT_VERSION, name
    )
```
the variable name is misleading since we've refactored things.

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+    # local cache.
+    index_exists = True
+    try:
+        BINARY_INDEX._fetch_and_cache_index(url_and_version)
```
```suggestion
        BINARY_INDEX._fetch_and_cache_index(metadata)
```

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+    if sources:
+        for source in sources:
+            tty.debug(f"reading specs from source: {source}")
+            hashes.extend(read_concrete_hashes(source))
```
Should we rename this option from `source(s)` to `env(s)`? That's the only thing it's allowed to be at this point.

kwryankrattiger replied:
I thought about it, but I didn't because `envs` has other implied behavior for other commands with `-e` so I left it as sources to avoid confusion.

I think we also want to add more options here, so I left the generic name for that option.

kwryankrattiger replied:
I thought about it, but I didn't because `envs` has other implied behavior for other commands with `-e` so I left it as sources to avoid confusion.

I think we also want to add more options here, so I left the generic name for that option.

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+        pass
+
+    if update_mode == ViewUpdateMode.APPEND:
+        tty.warn(
+            "Appending to a view index does not guarantee idempotent write when contending "
+            "with multiple writers. This feature is meant to be used by a single process."
+        )
```
I think this should be a tty.debug

1. if you see this warning and you are using multiple processes, it's a little late to do anything about it anyway.
2. we've pretty heavily documented that this isn't process safe
3. a warning is pretty loud for something we expect people to do regularly when they know what they're doing

kwryankrattiger replied:
> if you see this warning and you are using multiple processes, it's a little late to do anything about it anyway.

We could add a `-y/--yes-to-all` option here. Also, most real cases people will have at least a few seconds before this actually does anything destructive to the build cache. Plenty of time to reflect on their actions.

> we've pretty heavily documented that this isn't process safe

This is one of those things I don't think we can over document. Spack is, in pretty much all other cases, able to properly do locking for this type of action. For people who don't read the docs (which is most people) they should at least get some kind of warning. imo this is minimal noise to ensure people aren't surprised by known issues.

> a warning is pretty loud for something we expect people to do regularly when they know what they're doing

I think we warning about a lot of stuff that people regularly do, but are known to actively or eventually cause issues for the user, whether they can do anything about it or not.

ie.
```
==> Warning: Spack is trying attach a Python dependency to 'meson@=0.53.2 build_system=python_pip platform=linux os=ubuntu22.04 target=x86_64'. This feature is deprecated, and will be removed in v1.2. Please make the dependency explicit in your configuration.
```

becker33 replied:
We may be overwarning, but I don't think we should warn on the common use-case as a general rule. We can fix that in other places as we find it.

I think a `-y/--yes-to-all` option and prompting for interaction is a reasonable option here.

kwryankrattiger replied:
Dropped the error when `-y` is passed. I think that is a fair compromise since I think that says pretty clearly the user is fine with the consequences.

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+    mirror = args.mirror
+    verify = set()
+    for v in args.verify:
+        verify.update(v)
```
```suggestion
    verify = set(args.verify)
```

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+        pass
+
+    # Check if the index exists, and cache it locally for next operations
+    url_and_version = spack.binary_distribution.MirrorMetadata(
```
We should move off these old names now that they carry more information. Otherwise this code will be very hard to maintain for anyone who wasn't part of this discussion.

becker33 commented on lib/spack/spack/test/binary_distribution.py:
```diff
@@ -1362,3 +1389,39 @@ def test_get_entries_from_cache_nested_mirrors(monkeypatch, tmp_path: pathlib.Pa
     assert len(spec_manifests_nested) == 1
+
+
+def test_url_and_version():
```
```suggestion
def test_mirror_metadata():
```

becker33 commented on lib/spack/spack/test/binary_distribution.py:
```diff
@@ -1362,3 +1389,39 @@ def test_get_entries_from_cache_nested_mirrors(monkeypatch, tmp_path: pathlib.Pa
+        spack.binary_distribution.MirrorMetadata.from_string("https://dummy.io/__v3@@4")
+
+
+def test_url_and_version_with_view():
```
```suggestion
def test_mirror_metadata_with_view():
```

becker33 commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -790,18 +839,245 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
+        mirror_update_keys(mirror)
+
+
+def check_index_fn(args):
```
I think it would make sense to push most of the content of this method into the `spack.mirror` module and leave the output and argument handling here. That way tests and other internal code can use this check as well.

But I'm ok with leaving that as TODO for a next PR to get this in.

kwryankrattiger replied:
Thanks. Yeah I think the relationship between `cmd.mirror` and `cmd.buildcache` needs work in general. It is far too confusing.

kwryankrattiger replied:
Thanks. Yeah I think the relationship between `cmd.mirror` and `cmd.buildcache` needs work in general. It is far too confusing.

kwryankrattiger committed:
Address review comments

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Fix test post environment defined views

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

becker33 commented on lib/spack/spack/test/binary_distribution.py:
```diff
@@ -292,6 +294,39 @@ def test_use_bin_index(monkeypatch, tmp_path: pathlib.Path, mutable_config):
+        install_cmd("--add", "--fake", "--no-cache", s.name)
+        buildcache_cmd("push", "-u", "test", s.name)
+
+    buildcache_cmd("update-index", "test", "testenv")
```
We should also test this omitting the env name and relying on it being the active environment.

kwryankrattiger replied:
Sure we can parameterize this test like the others.

kwryankrattiger replied:
actually, parametrization is going to be a pain here. I will just copy it.

becker33 replied:
If you wanted to parametrize, something like:

```
context = ev.read("testenv") if use_active_env else nullcontext
args = ["update_index", "test"]
if not use_active_env:
    args.append("testenv")
with context:
    buildcache_cmd(*args)
```
should work, but I'm fine with separate tests as well.

kwryankrattiger committed:
Drop wrapper since there is only one type of source now

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Cleanup error messages

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+
+
+@contextlib.contextmanager
+def spec_lockfile(spec: spack.spec.Spec, prefix: str):
```
This doesn't actually use the env from the lockfile, it just uses an environment. We no longer have a special code path for a lockfile instead of an environment (because you can just create one from the lockfile if you don't have the environment for it anymore).

kwryankrattiger replied:
This is passing the path for the environment vs the name of the environment

kwryankrattiger replied:
ok. So I see your point but I think it is going to be better to leave it the way it is, ie. parameterized on each of these callbacks.

- having this always result in an active environment context is going to taint the tests where I explicitly want to _not_ include the active environment as a possible factor.
- whenever we come back to this to extend the unit tests here to cover new types of sources I don't want to have to completely refactor the tests to handle that, or copy-paste yet again.
- I would like to avoid as much logic for handling parametrization in the test as possible so that each of the cases is handle as simply has possible.
- The callbacks are quite simple, but I added additional docstring comments to express what they are meant to do and how they are being used.

Your other comments made me realize I was only testing this by passing the path to an environment, not the name. So I actually added yet another parametrization to exhaustively cover that case as well.

I also added another missing case (that this made me think of) where we need to ensure that passing an explicit path overrides the active environment.

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+        pass
+    yield [ev.environment_dir_from_name(env_name, exists_ok=True)]
+
+
```
These methods don't seem to warrant being separated -- I'll put code suggestions in the test that uses them showing what I mean.

Instead, we should have a helper method:

```
def create_test_env(name: str, spec: spack.spec.Spec) -> ev.Environment:
    with ev.create(name) as e:
        e.add(spec)
        e.concretize()
        e.write()
        return e
```

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+@pytest.mark.parametrize("source", (spec_env, spec_lockfile))
+def test_buildcache_create_view(
+    tmp_path, mutable_config, mutable_database, mutable_mock_env_path, source
+):
```
```suggestion
@pytest.mark.parametrize("by_name", (True, False))
def test_buildcache_create_view(
    tmp_path, mutable_config, mutable_database, mutable_mock_env_path, by_name
):
```

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+        hashes_in_view = read_specs_in_index(mirror_directory, "test_view")
+
+    # Create a view using a parametrized source that contains one of the mpileaks
+    with source(mpileaks_specs[0], tmp_path / "source_stage") as source_args:
+        command_args = ["update-index", "--name", "test_view", "my-mirror"]
+        if source_args:
+            command_args.extend(source_args)
+        buildcache(*command_args)
```
```suggestion
    # Create a view that contains one of the mpileaks
    with create_test_env("test_env", mpileaks_specs[0]) as e:
        command_args = ["update-index", "--name", "test_view", "my-mirror"]
        if by_name:
            command_args.extend("test-env")
        buildcache(*command_args)
```
If I'm reading correctly, this covers everything done by `spec_env` and `spec_lockfile` currently, but is much easier to read and understand what's going on.

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+
+    # Grab all of the hashes for each mpileaks spec
+    mpileaks_0_hashes = [s.dag_hash() for s in mpileaks_specs[0].traverse()]
+    mpileaks_1_hashes = [s.dag_hash() for s in mpileaks_specs[1].traverse()]
```
```suggestion
    mpileaks_0_hashes = {s.dag_hash() for s in mpileaks_specs[0].traverse()}
    mpileaks_1_hashes = {s.dag_hash() for s in mpileaks_specs[1].traverse()}
```

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+    for h in mpileaks_0_hashes:
+        assert h in hashes_in_view
+    for h in mpileaks_1_hashes:
+        assert h not in hashes_in_view or h in mpileaks_0_hashes
```
```suggestion
    for h in mpileaks_1_hashes - mpileaks_0_hashes:
        assert h not in hashes_in_view
```

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+        command_args = ["update-index", "-y", "--append", "--name", "test_view", "my-mirror"]
+        if source_args:
+            command_args.extend(source_args)
+        buildcache(*command_args)
```
```suggestion
    with create_test_env("test_env_1", mpileaks_specs[0]) as e:
        command_args = ["update-index", "-y", "--append", "--name", "test_view", "my-mirror"]
        if by_name:
            command_args.extend("test_env_1")
        buildcache(*command_args)
```

kwryankrattiger replied:
This isn't quite right, since there is an active environment in this context. I want to ensure that the passed path or name is they driver and it isn't being triggered by a side effect of the active environment.

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+        command_args = ["update-index", "-y", "--append", "--name", "test_view", "my-mirror"]
+        if source_args:
+            command_args.extend(source_args)
+        buildcache(*command_args)
```
Same as above

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+
+    # Grab all of the hashes for each mpileaks spec
+    mpileaks_0_hashes = [s.dag_hash() for s in mpileaks_specs[0].traverse()]
+    mpileaks_1_hashes = [s.dag_hash() for s in mpileaks_specs[1].traverse()]
```
```suggestion
    mpileaks_0_hashes = {s.dag_hash() for s in mpileaks_specs[0].traverse()}
    mpileaks_1_hashes = {s.dag_hash() for s in mpileaks_specs[1].traverse()}
```

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +977,231 @@ def mock_stat_url(url: str):
+
+
+@pytest.mark.parametrize("source", (spec_env, spec_lockfile))
+def test_buildcache_create_view_overwrite(
```
Same changes for this test as the previous two

becker33 reviewed (changes requested):
The tests still need work to be easier to read for future maintainers who weren't part of this conversation. Currently they're very influenced by the complexity of the original design vs the current code.

kwryankrattiger committed:
check_index support passing view name override

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
mirror create supporting creating with view name

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Add more testing, better parameter names

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Try intalling zig?

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Drop zig install, maybe this was ephemeral

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

becker33 commented on lib/spack/spack/test/binary_distribution.py:
```diff
@@ -292,6 +294,71 @@ def test_use_bin_index(monkeypatch, tmp_path: pathlib.Path, mutable_config):
+    with ev.read("testenv"):
+        install_cmd("--add", "--fake", "--no-cache", s.name)
+        buildcache_cmd("push", "-u", "test", s.name)
+        buildcache_cmd("update-index", "test", "testenv")
```
```suggestion
        buildcache_cmd("update-index", "test")
```
We should be testing relying on the active env, not passing the name.

kwryankrattiger committed:
Test refactor

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +975,346 @@ def mock_stat_url(url: str):
+
+
+def args_for_active_env(spec: spack.spec.Spec):
+    """Build cache index view source is an environment path"""
```
```suggestion
    """Build cache index view source is an active environment"""
```

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +975,346 @@ def mock_stat_url(url: str):
+
+
+def args_for_env_by_path(spec: spack.spec.Spec):
+    """Build cache index view source is an managed environment name"""
```
```suggestion
    """Build cache index view source is an environment path"""
```

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +975,346 @@ def mock_stat_url(url: str):
+def args_for_env_by_path(spec: spack.spec.Spec):
+    """Build cache index view source is an managed environment name"""
+    env = create_env_from_concrete_spec(spec)
+    return [nullcontext(), [env.name]]
```
```suggestion
    return [nullcontext(), [env.path]]
```

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +975,346 @@ def mock_stat_url(url: str):
+
+
+def args_for_env_by_name(spec: spack.spec.Spec):
+    """Build cache index view source is an managed environment name"""
```
```suggestion
    """Build cache index view source is a managed environment name"""
```

becker33 commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -976,3 +975,346 @@ def mock_stat_url(url: str):
+
+    hashes_in_view = read_specs_in_index(mirror_directory, "test_view")
+    # Assert all of the hashes for mpileaks_0_hashes exist in the view, and no other hashes
+    assert not (hashes_in_view - mpileaks_0_hashes)
```
```suggestion
    assert hashes_in_view == mpileaks_0_hashes
```

becker33 reviewed (changes requested):
Minor copy/paste issues in the new code, otherwise LGTM

kwryankrattiger committed:
make force and append option mutually exclusive

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
Add clarification in docs and fix some grammar

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>

kwryankrattiger committed:
RTD title underline fix up

Signed-off-by: Ryan Krattiger <ryan.krattiger@kitware.com>
